doc/update.md

```diff
@@ -0,0 +1,94 @@
+# Xi view update protocol
+
+This document describes a proposal for a new protocol for sending view updates from the core to the front-end.
+
+## Background
+
+Goals: keep everything async. Keep network traffic minimal. Allow front-end to retain as much information as possible (including text if only cursors are updated). Allow front-end to use small amounts of memory even when document is large.
+
+Conceptually, the core maintains a full view of the document, which can be considered an array of lines. Each line consists of the *text* (a string), a set of *cursor* locations, and a structure representing *style* information. Many operations update this view, at which point the core sends a notification to the front-end.
+
+The front-end maintains a *cache* of this view. Some lines will be present, others will be missing. A cache is consistent with the true state when all present lines match.
+
+To optimize communication, the core keeps some state about the client. One bit of this state is the *scroll window;* in general, the core tries to proactively update all lines within this window (plus a certain amount of slop on top and bottom). In addition, the core maintains a set of lines in the client's cache. If a line changes, the update need only be communicated if it is in this set. This set is conservative; if a line is missing in the actual cache held by the front-end (evicted to save memory), no great harm is done updating it.
+
+## Requests from front-end to core
+
+```
+scroll: [number, number]  // first line, last line
+```
+
+```
+request: [number, number]  // first line, last line
+```
+
+## Requests from core to front-end
+
+
+```
+set_style
+  id: number
+  fg_color?: number // 32-bit RGBA value
+  bg_color?: number // 32-bit RGBA value, default 0
+  weight?: number // 100..900, default 400
+  italic?: boolean  // default false
+```
+
+It's not hard to imagine more style properties (typeface, size, OpenType features, etc).
+
+The guarantee on id is that it is not currently in use in any lines in the view. However, in practice, it will probably just count up. It can also be assumed to be small, so using it as an index into a dense array is reasonable.
```

Comment:
nit: this section could say that id will be used to reference a style when updates are sent, right now it's only explicitly stated at the very bottom of the document.